test/framework/managedcluster.go

```diff
@@ -0,0 +1,190 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/gomega"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+func GetManagedCluster(hub *Hub, clusterName string) (*clusterv1.ManagedCluster, error) {
```

Comment:
If these are all hub side operation, shouldn't we make it as func in hub struct?

Reply:
That's a good question, I think a lot on this, whether the `GetManagedCluster` and other similar code blocks be a method of hub or a function.

The reason 1 is to keep the defination and scope of a `hub` clear, hub is only:
* the client
* the metadata
* the runtime data
It provide information for us to do things, but not do things for us.

The reason 2 is,  the `GetManagedCluster` actually doesn't need the whole hub, it only needs the `clusterclient`,  and if we have:

```go
func GetManagedCluster(clusterclient clusterclient.Interface, clusterName string) (*clusterv1.ManagedCluster, error) {
	if clusterName == "" {
		return nil, fmt.Errorf("the name of managedcluster should not be null")
	}
	return clusterclient.ClusterV1().ManagedClusters().Get(context.TODO(), clusterName, metav1.GetOptions{})
}
```

That means we can reuse this function and this framework not only in e2e, but also the intergration test, the addon repo's e2e and integration, the downstream repos that replies on ocm.

Currently if we develop a addon or a component that relaies on ocm, we write integration and e2e part(code and scripts) indenependly but if we can provide a ocm-test-framework, I think it can reduce some effort in write and maintaining these tests.

Reply:
To my undstanding, hub defines a hub environment. A test case could manage hub environment with the hub struct.

Reply:
Sure, I can change to `hub.GetManagedCluster` since currently we have no plan for a ultimate test-framework for al l ocm related repos. But I still think `GetManagedCluster` is not limit to a hub environment, in integration cases, there is not hub just controllers and fake kube-apiserver, we can still reuse `GetManagedCluster`.

Reply:
Updated, all functions are moved under hub and spoke as method.